### PR TITLE
Add blog about ephemeral container security

### PR DESCRIPTION
**Pull request by phillebaba**

### CODE REVIEW

File: blog/2022-04-12-ephemeral-containers.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+  - kubernetes
+  - security
+  - ephemeral container
+---
+
+Attempting to debug a Pod and realizing that you can't install curl due to security settings has to be a meme at this point. Good security practices is always nice but it often comes at the cost of usability. To the point where some may even solve this problem by installing debug tools into their production images. Shudders.
+
+<img src="https://i.imgflip.com/6cczqi.jpg" title="made at imgflip.com"/>
+
+Kubernetes has introduced a new concept called [ephemeral containers](https://kubernetes.io/docs/concepts/workloads/pods/ephemeral-containers/) to deal with this problem. Ephemeral containers are temporary containers that can be attached after a Pod has been created. Rejoice! We can now attach a temporary container will all the tools which we desire. While the applications container may have "annoying security features" like a read only file system the ephemeral container can enjoy all the freedom which writing files entails. I love this feature so I need to upgrade my cluster immediately!
```
Comment (simongottschlag) [inline blog/2022-04-12-ephemeral-containers.md:19]: s/will all/with all

**Review by simongottschlag — DISMISSED** [review]

**Comment by simongottschlag** [comment]
@phillebaba good job! Only had a small comment.

Did you find any information about RBAC and ephemeralContainers? It would be nice with a few lines about it

**Review by NissesSenap — COMMENTED** [review]
There is a way of having an expander that only show the top of the blog post. Look at @bittrance post about 12 factor application. There you will see a example of this

File: blog/2022-04-12-ephemeral-containers.md [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+### Pod Security Policies
+
+[Pod Security Policies](https://kubernetes.io/docs/concepts/security/pod-security-policy/) used to be the default policy tool for Kubernetes, and a lot of projects have rules based on Pod Security Policies (PSP). However if you are relying on PSP in a modern cluster you should really start looking for other options like OPA Gatekeeper or Kyverno. PSP has been deprecated since Kubernetes v1.21 and will be removed in v1.25.
+
+If PSP is your only policy tool and you are planning to upgrade to v1.23, don't. As PSP is deprecated no new features have been added, and that includes policy enforcement on ephemeral containers. Which means that any security context in an ephemeral container is allowed no matter the PSP in the cluster.
```
Comment (NissesSenap) [inline blog/2022-04-12-ephemeral-containers.md:65]: Would it be good to mention the new version of psp is coming? https://kubernetes.io/docs/concepts/security/pod-security-admission/

**Review by NissesSenap — APPROVED** [review]
Great job! 

Not for this blog but In a new one I'm wondering how debug containers work In  a gitops setting?
I would assume flux would remove it.
Should you add some exclusion of looking for ephemeral containers or is it better to keep it etc.

File: blog/2022-04-12-ephemeral-containers.md (hunk 2 above)
Comment (phillebaba) [reply blog/2022-04-12-ephemeral-containers.md:65]: Was not aware of that work. Is it worth mentioning if I havent verified that it actually blocks?

**Comment by phillebaba** [comment]
@NissesSenap good question regarding Flux. It should not have any affect as long as you do not create a Pod directly. If Flux creates a deployment it will not detect any changes as the ephemeral container is added to the Pod.

File: blog/2022-04-12-ephemeral-containers.md (hunk 2 above)
Comment (NissesSenap) [reply blog/2022-04-12-ephemeral-containers.md:65]: I think it's enough to just mention that it's coming and if you are a psp user it's something that is worth looking into.

File: blog/2022-04-12-ephemeral-containers.md [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+  which have to be resolved before it can be enabled.
+authors:
+  - name: Philip Laine
+tags:
+  - kubernetes
+  - security
+  - ephemeral container
+---
+
+Attempting to debug a Pod and realizing that you can't install curl due to security settings has to be a meme at this point. Good security practices is always nice but it often comes at the cost of usability. To the point where some may even solve this problem by installing debug tools into their production images. Shudders.
```
Comment (bittrance) [inline blog/2022-04-12-ephemeral-containers.md:15]: "practices *are* always nice"

File: blog/2022-04-12-ephemeral-containers.md [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+<!-- truncate -->
+
+<img src="https://i.imgflip.com/6cczqi.jpg" title="made at imgflip.com"/>
+
+Kubernetes has introduced a new concept called [ephemeral containers](https://kubernetes.io/docs/concepts/workloads/pods/ephemeral-containers/) to deal with this problem. Ephemeral containers are temporary containers that can be attached after a Pod has been created. Rejoice! We can now attach a temporary container with all the tools which we desire. While the applications container may have "annoying security features" like a read only file system the ephemeral container can enjoy all the freedom which writing files entails. I love this feature so I need to upgrade my cluster immediately!
+
+## Digging Deeper
+
+Now that we have the new feature we can start a ephemeral container in what ever Pod we like.
```
Comment (bittrance) [inline blog/2022-04-12-ephemeral-containers.md:25]: "whatever" though I would personally prefer "in any Pod".

**Review by NissesSenap — APPROVED** [review]

File: blog/2022-04-12-ephemeral-containers.md [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+  - name: debugger-r59b7
+    image: busybox:1.28
+    imagePullPolicy: IfNotPresent
+    stdin: true
+    terminationMessagePath: /dev/termination-log
+    terminationMessagePolicy: File
+    tty: true
+```
+
+Interesting, there is a new field called `ephemeralContainers` in the Pod. This new field contains a list of containers similar to `initContainers` and `containers`. It is not identical as there are certain options which are not available, refer to the [API documentation](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#ephemeralcontainer-v1-core) for more information. It does however allow configuration of the container security context, which could in theory allow a bad actor to escalate the containers privileges. This should not affect those of us who use a policy enforcement tool right? The answer is yes and no depending on the tool and version that is being used. It also depends on if you are using policies from the projects library or policies developed in house.
```
Comment (bittrance) [inline blog/2022-04-12-ephemeral-containers.md:51]: "in the Pod definition."
"the container's privileges"
"the project's library"

File: blog/2022-04-12-ephemeral-containers.md [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+> Cluster administrators will be expected to choose from one of the following mechanisms for restricting usage of ephemeral containers:
+> * Use RBAC to control which users are allowed to access the /ephemeralcontainers subresource.
+> * Write or use a third-party admission controller to allow or reject Pod updates that modify ephemeral containers based on the content of the update.
+> * Disable the feature using the EphemeralContainers feature gate.
+
+RBAC is additive which means that it is not possible to remove permissions from a role. This type of mitigation obviously does not matter if all users a cluster admin, which they should not be, so we assume that new roles are created for the cluster consumers. In this case having a look at the existing roles can be enough to make sure that the subresource is not included in the role.
```
Comment (bittrance) [inline blog/2022-04-12-ephemeral-containers.md:98]: Name the sub-resource to exclude?

File: blog/2022-04-12-ephemeral-containers.md [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+  - create
+  - delete
+  - deletecollection
+  - patch
+  - update
+```
+
+## Checking Policy Enforcement
+
+So what if we have a cluster with ephemeral containers enabled, and you are unsure if the correct policies are enforced? Or you just want to verify you work. The debug command does not expose any options to set any security context configuration, so we need another option. Ephemeral containers cannot be defined in a Pod when it is created neither can it be added with an update. The Kubernetes documentation states the following.
```
Comment (bittrance) [inline blog/2022-04-12-ephemeral-containers.md:124]: Not sure what you want to say with "verify you work"? Should it be "verify that ephemeral containers work"?
"when it is created *and* neither can it ..."

File: blog/2022-04-12-ephemeral-containers.md [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+}
+```
+
+Run the program and pass the namespace, pod name, and path to a kube config file. We assume that the ephemeral-demo Pod is still running. 
+
+```shell
+go run main.go default ephemeral-demo $KUBECONFIG
+```
+
+If it completes with no error a privileged ephemeral container should have been added to the Pod. Exec into it and list the hosts devices to prove that it is a privileged container.
```
Comment (bittrance) [inline blog/2022-04-12-ephemeral-containers.md:204]: "list the host's devices"

File: blog/2022-04-12-ephemeral-containers.md [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+
+```shell
+kubectl exec -it ephemeral-demo -c debug -- sh
+ls /dev
+```
+
+## Conclusion
+
+If there is one thing that you learn from this post, it is that any policy tool that has not been updated in the last couple of months will not enforce rules on ephemeral containers. Additionally you should really analyze any in house developed policies to include checking ephemeral containers.
```
Comment (bittrance) [inline blog/2022-04-12-ephemeral-containers.md:213]: I would prefer "one takeaway from this post".

File: blog/2022-04-12-ephemeral-containers.md [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+
+Some may argue that this type of oversight is not really an issue. Ephemeral containers can't mount host paths, share host PID, or host networking. All it can do is set the common container security context. That is a fair comment, because it's true. Being able to create a privileged container is however still not ideal, and there are [methods to escalate privileges](https://bishopfox.com/blog/kubernetes-pod-privilege-escalation#Pod3) when this is possible. Either way it is important to be aware of how policies are enforced and the security contexts which are allowed.
```
Comment (bittrance) [inline blog/2022-04-12-ephemeral-containers.md:215]: What does "share the host PID" mean?

File: blog/2022-04-12-ephemeral-containers.md [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+
+I am still not sure how much of an issue this will be short term. Cloud providers are currently in the process of rolling out Kubernetes v1.23 in their SaaS offerings. In these solutions it is still a possibility that they chose to disable ephemeral containers. Those rolling their own clusters may have already upgraded to v1.23 and not be aware of the new feature. That is the biggest issue really, that the end user has to be aware of the existence of ephemeral containers. A company may have invested in a security audit 6 months ago but that is only valid as long as the same Kubernetes version is used. The reality is that these types of things are easy to miss, especially when one assumes that the policy tool just solves the issue without verifying that it does.
```
Comment (bittrance) [inline blog/2022-04-12-ephemeral-containers.md:217]: I would have used the phrase "[not] secure by default" somewhere here.

**Review by bittrance — APPROVED** [review]
Looking good. Some minor comments.

**Review by bittrance — APPROVED** [review]

**Pull request merged**